### PR TITLE
feat: do not kill ec when ecm restart  part-1 (#4184)

### PR DESCRIPTION
**Pull request by CharlieYan24**
### What is the purpose of the change

1) When the ECM restarts, choose not to kill the engine. 
2) StatusUpdate message,  pid update message , yarn appid update message  of engine server(EM) directly send to linkis-manager service(LM), Make engine conn manager(ECM) service stateless .

 
### Related issues/PRs

Related issues: #4184 
Related pr:#4184

### Brief change log

- Replace the function of DefaultEngineConnListService class in Ecm with DefaultEngineConnManageService, DefaultEngineConnStatusCallbackService , DefaultEngineConnPidCallbackService.
(StatusUpdate message,  pidUpdate message , yarnAppidUpdate message  of engine conn(EM) service directly send to linkis-manager service(LM))
- Modify EMRestfulApi.executeECMOperation method add logDirSuffix parameter.
- Modify DefaultEMEngineService.stopEngine method parameter, adding  engineType, logDirSuffix.

### Checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [Linkis mailing list](https://linkis.apache.org/community/how-to-subscribe) first)
- [x] **If this is a code change**: I have written unit tests to fully verify the new behavior.

### CODE REVIEW

**Comment by CharlieYan24** [comment]
image
Replace the function of DefaultEngineConnListService class in Ecm with DefaultEngineConnManageService, DefaultEngineConnStatusCallbackService , DefaultEngineConnPidCallbackService.
(StatusUpdate message, pidUpdate message , yarnAppidUpdate message of engine conn(EM) service directly send to linkis-manager service(LM))

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/impl/DefaultEngineConnStatusCallbackService.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  public void dealEngineConnStatusCallback(EngineConnStatusCallback protocol) {
+    logger.info(
+        "EngineConnStatusCallbackServiceImpl handle engineConnStatus callback serviceInstance: [{}] status: [{}]",
+        protocol.serviceInstance(),
+        protocol.status());
+    if (!NodeStatus.isAvailable(protocol.status())) {
+      EngineStopRequest engineStopRequest = new EngineStopRequest();
+      engineStopRequest.setServiceInstance(protocol.serviceInstance());
+      engineStopRequest.setUser("hadoop");
+      engineStopService.stopEngine(
```
Comment (peacewong) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/impl/DefaultEngineConnStatusCallbackService.java:71]: Need to add try catch to ensure the execution of dealEngineConnStatusCallbackToAM

Comment (CharlieYan24) [reply linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/impl/DefaultEngineConnStatusCallbackService.java:71]: I have add try catch around stopEngine method to ensure the execution of dealEngineConnStatusCallbackToAM, thanks for your kind remind.

File: linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/conf/GovernanceCommonConf.scala [hunk 2]
```diff
@@ -43,6 +43,9 @@ object GovernanceCommonConf {
   val ENGINE_CONN_MANAGER_SPRING_NAME =
     CommonVars("wds.linkis.engineconn.manager.name", "linkis-cg-engineconnmanager")
 
+  val ENGINE_APPLICATION_MANAGER_SPRING_NAME =
```
[the remark this answers is not in the document]
Comment (CharlieYan24) [reply linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/conf/GovernanceCommonConf.scala:46]: Already changed MANAGER_SERVICE_NAME 's default value to ENGINE_APPLICATION_MANAGER_SPRING_NAME.

File: linkis-computation-governance/linkis-engineconn/linkis-computation-engineconn/src/test/scala/org/apache/linkis/engineconn/computation/executor/upstream/access/ECTaskEntranceInfoAccessHelper.scala [hunk 3]
```diff
@@ -53,7 +53,10 @@ object ECTaskEntranceInfoAccessHelper {
     val host = CommonVars(Environment.ECM_HOST.toString, "127.0.0.1").getValue
     val port = CommonVars(Environment.ECM_PORT.toString, "80").getValue
     engineCreationContext.setEMInstance(
```
[the remark this answers is not in the document]
Comment (CharlieYan24) [reply linkis-computation-governance/linkis-engineconn/linkis-computation-engineconn/src/test/scala/org/apache/linkis/engineconn/computation/executor/upstream/access/ECTaskEntranceInfoAccessHelper.scala:55]: Yes, I have removed this unused file.

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/scala/org/apache/linkis/manager/am/service/em/DefaultEMEngineService.scala [hunk 4]
```diff
@@ -95,6 +100,17 @@ class DefaultEMEngineService extends EMEngineService with Logging {
     )
     val engineStopRequest = new EngineStopRequest
     engineStopRequest.setServiceInstance(engineNode.getServiceInstance)
+
+    val ecInfo: ECResourceInfoRecord =
+      ecResourceInfoService.getECResourceInfoRecordByInstance(
+        engineNode.getServiceInstance.getInstance
+      )
+    // append engineType and logDirSuffix
+    if (ecInfo != null) {
```
[the remark this answers is not in the document]
Comment (CharlieYan24) [reply linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/scala/org/apache/linkis/manager/am/service/em/DefaultEMEngineService.scala:109]: Thanks for your detailed guidance .I have refactored the code following  above guidance. Thanks.

File: linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/conf/GovernanceCommonConf.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (peacewong) [reply linkis-computation-governance/linkis-computation-governance-common/src/main/scala/org/apache/linkis/governance/common/conf/GovernanceCommonConf.scala:46]: > Already changed MANAGER_SERVICE_NAME 's default value to ENGINE_APPLICATION_MANAGER_SPRING_NAME.
Thanks!

File: linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/DefaultEngineConnKillService.java [hunk 5]
```diff
@@ -220,6 +187,20 @@ private boolean isProcessAlive(String pid) {
         }
     }
 
+    private String findProcessId(String processPort){
+        String findCmd = "lsof -t -i:" + processPort;
+        List<String> cmdList = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (peacewong) [reply linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/DefaultEngineConnKillService.java:192]: > Already use sudo and move this method to GovernanceUtils, Thanks for your good advice.

Thank you!

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/impl/DefaultEngineConnStatusCallbackService.java (hunk 1 above)
Comment (peacewong) [reply linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/impl/DefaultEngineConnStatusCallbackService.java:71]: > I have add try catch around stopEngine method to ensure the execution of dealEngineConnStatusCallbackToAM, thanks for your kind remind. Thanks for your time.

Thanks!

**Comment by peacewong** [comment]
These new codes are not pushed? @CharlieYan24

**Comment by CharlieYan24** [comment]
> These new codes are not pushed? @CharlieYan24

I was trapped in building test environment and test the above code those few days. And fixed the serialize problem  in /userResources  and /getApplicationList api in RMMonitorRest class.

**Review by peacewong — APPROVED** [review]
LGTM.

**Review by peacewong — APPROVED** [review]

**Pull request merged**